### ISSUE TITLE
Must we require ESMFMKFILE? Can we have a FindESMF.cmake which does this for us?

### ISSUE THREAD

**Issue by edwardhartnett** [NOAA-EMC/UFS_UTILS#358]
If I try to build without setting ESMFMKFILE, I get this error:

```
  ESMFMKFILE not defined.  This is the path to esmf.mk file.  Without this
  filepath, ESMF_FOUND will always be FALSE.
```

This is yucky, as I'm sure we all agree. Efforts to get ESMF to use a better build system are underway (and have been for over a decade, so that's a goal I may not achieve in the near-term).

But, even if I set CMAKE_PREFIX_PATH to include the directory with esmf.mk, I still get the yucky error.

Can we have a FindESMF.cmake, which can set ESMFMKFILE, after finding ESMF from the CMAKE_PREFIX_PATH?

That way, we would hide this yuckiness.

@aerorahul can you help? I don't know how to do this, but it seems like the sort of thing cmake is good at...

**Comment by aerorahul** [comment]
One can edit `FindESMF.cmake` to query `esmf_ROOT` or `esmf_DIR` and set `ESMFMKFILE` internally based on those 2 standard CMake variables.
That is just the same as any other `FindPackage.cmake` file that relies on `package_ROOT` or `package_DIR` variables.

You still won't be able to use `CMAKE_PREFIX_PATH`.  `CMAKE_PREFIX_PATH` is only usable if a package is built with cmake; i.e. it has cmake generated config file.  ESMF is not CMake built.  `FindESMF.cmake` is written by hand to extract information from `esmf.mk` and create the linker and include lines that would have been obtained by directly sourcing `esmf.mk`.

You can however append to `CMAKE_MODULE_PATH` the path to `FindESMF.cmake`, which is what we do.

In short, `FindESMF.cmake` allows us to do this:
```
target_link_libraries(libname esmf)
```

instead of

```
target_link_libraries(libname {ESMF_F90ESMFLINKRPATHS} ${ESMF_F90ESMFLINKPATHS} ${ESMF_F90LINKPATHS} ${ESMF_F90LINKLIBS} ${ESMF_F90LINKOPTS}" ESMF_INTERFACE_LINK_LIBRARIES))
```
and possibly a lot more for includes.

Does that help.

**Comment by edwardhartnett** [comment]
What I want is to not have to set a special environment variable for ESMF. But apparently I will still have to, since it has to be set somewhere, and CMAKE_PREFIX_PATH will not work.

**Comment by aerorahul** [comment]
Like I said, `CMAKE_PREFIX_PATH` does not work because `ESMF` is not a built with cmake and is not providing a `esmf-config.cmake` file.  If it were, one wouldn't need to set any special variables and `CMAKE_PREFIX_PATH` would work.  One could manually write a `esmf-config.cmake` file and emulate `FindESMF.cmake` in it and its associated `esmf-targets.cmake`, `esmf-config-version.cmake` etc. 

This is the same reason for the existence of  other `FindPackage.cmake` files that we may have written or are using
 e.g. `FindNetCDF.cmake`.  If netCDF was built with cmake and was properly providing its targets and dependencies, one would not need to write the `FindNetCDF.cmake` and simply use `netcdf-config.cmake` by providing `CMAKE_PREFIX_PATH` to the netCDF installation directory.

I don't see a way out of specifying some kind of environment variable for ESMF.  Atleast `esmf_ROOT` is a standard (in some ways) than `ESMFMKFILE`.

**Comment by edwardhartnett** [comment]
What if we wrote a cmake wrapper for the ESMF install? Would that work?

**Comment by aerorahul** [comment]
I don't think so, but I could be wrong.  The `targets` are cmake targets and the dependencies are created and resolved by cmake.  Wrapping them up, if I understand what you are suggesting, is basically calling the underlying build via cmake.
The targets themselves need to be created via `add_library` calls, and the package has to be configured using cmake package configurator.

Besides, the underlying build is vast for the numerous platforms, compilers, options, etc it supports.

**Comment by edwardhartnett** [comment]
Yuck.

OK I will ponder and await illumination. Thanks for the info, I will study your answers...
